### ISSUE TITLE
TypeError: expected string or bytes-like object

### ISSUE THREAD

**Issue by ryandesign** [mypaint/mypaint#1095]
#### Description of the problem

Hi, I'm the maintainer of [MyPaint in MacPorts](https://ports.macports.org/port/MyPaint). I don't remember seeing this when I first updated MyPaint to 2.0.1, but now, when I quit MyPaint, it displays a Python traceback (below). Probably some of MyPaint's dependencies have been updated to newer versions in the mean time; perhaps that's the cause. For one thing, we were using py-gobject3 3.28.3 before and it was updated to 3.36.1. This has allowed me to switch MyPaint to Python 3.8 (but I see the same backtrace now even if I stick with Python 3.7).

#### Basic system details

MyPaint version: 2.0.1
Operating system: macOS 10.13.6
Desktop environment: ?

#### Steps to reproduce

1. Start MyPaint
2. Quit MyPaint

#### Backtraces or error messages

```
INFO: mypaint: There is an old-style configuration area in '/Users/rschmidt/.mypaint/'
INFO: mypaint: Its contents can be migrated to $XDG_CONFIG_HOME and $XDG_DATA_HOME if you wish.
INFO: mypaint: See the XDG Base Directory Specification for info.
INFO: gui.main: No locale setting found, using system locale
INFO: lib.i18n: OSX: imported AppKit.NSLocale OK. Will use for LANG, and for LANGUAGE order.
INFO: lib.i18n: OSX: LANG='en_US.UTF-8'
INFO: lib.i18n: OSX: LANGUAGE='en_US'
INFO: lib.i18n: Value of LANGUAGE after cleanup: 'en_US:C'
INFO: gui.compatibility: Setting mode to 2.x (standard)
INFO: gui.compatibility: Setting default layer type to Pigment
INFO: gui.device: New device 'eraser' (GDK_SOURCE_ERASER, axes:5, class=X11DeviceXI2, vendor=None, product=None)
INFO: gui.device: New device 'cursor' (GDK_SOURCE_CURSOR, axes:5, class=X11DeviceXI2, vendor=None, product=None)
INFO: gui.device: New device 'pen' (GDK_SOURCE_PEN, axes:5, class=X11DeviceXI2, vendor=None, product=None)
INFO: gui.device: New device 'xquartz virtual pointer' (GDK_SOURCE_MOUSE, axes:6, class=X11DeviceXI2, vendor=None, product=None)
INFO: gui.device: New device 'Virtual core XTEST pointer' (GDK_SOURCE_MOUSE, axes:2, class=X11DeviceXI2, vendor=None, product=None)
INFO: gui.document: Initialized background from '/opt/local/Library/Frameworks/Python.framework/Versions/3.8/share/mypaint/backgrounds/mrmamurk/mamurk_e_1.png'
WARNING: gui.keyboard: Ignoring keybinding for '<Actions>/BrushModifierActions/BlendModeMenu'
INFO: gui.brushmanager: Switching default pigment setting to On
INFO: lib.document: load_ora: '/Users/rschmidt/.mypaint/scratchpads/autosave.ora'
ERROR: lib.layer.tree: Could not load any layer, document is empty.
INFO: lib.layer.tree: Adding an empty painting layer
INFO: lib.document: 0.163s load_ora total
INFO: gui.filehandling: Loaded scratchpad from '/Users/rschmidt/.mypaint/scratchpads/autosave.ora'
INFO: gui.compatibility: Setting default layer type to Pigment
--- Logging error ---
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/logging/__init__.py", line 1081, in emit
    msg = self.format(record)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/logging/__init__.py", line 925, in format
    return fmt.format(record)
  File "/opt/local/bin/mypaint", line 84, in format
    msg = token_re.sub(repl, msg)
TypeError: expected string or bytes-like object
Call stack:
  File "/opt/local/bin/mypaint", line 309, in <module>
    main.main(
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/main.py", line 231, in main
    run()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/main.py", line 224, in run
    Gtk.main()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/gi/overrides/Gtk.py", line 1644, in main
    return _Gtk_main(*args, **kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/drawwindow.py", line 678, in quit_cb
    self.app.save_gui_config()  # FIXME: should do this periodically
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/application.py", line 612, in save_gui_config
    self.save_settings()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/application.py", line 493, in save_settings
    self.brushmanager.save_brush_history()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/brushmanager.py", line 1077, in save_brush_history
    brush.save()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/brushmanager.py", line 1449, in save
    brushinfo = self.brushinfo.clone()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/brushmanager.py", line 1304, in get_brushinfo
    self._ensure_settings_loaded()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/brushmanager.py", line 1555, in _ensure_settings_loaded
    self.load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/brushmanager.py", line 1496, in load
    self._load_settings()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/gui/brushmanager.py", line 1519, in _load_settings
    self._brushinfo.load_from_string(brushinfo_str)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/lib/brush.py", line 380, in load_from_string
    self._load_old_format(settings_str)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/mypaint/lib/brush.py", line 468, in _load_old_format
    logger.warning(error)
Message: ('parent_brush_name deevad%2Fkneaded', 'Cann')
Arguments: ()
```

**Comment by jplloyd** [comment]
Thanks for the report.

The error is from the logging hook used to add color to the stdout/stderr messages, but it looks like the root cause comes down to some Py3 compatibility that is either no longer valid, or was never valid to begin with. This code path is only hit when loading brushes with the old-style (non-json) format - a change that was made ~8 years ago.

At any rate, shouldn't be hard to fix, and it's not related to pygobject.

**Comment by ryandesign** [comment]
> This code path is only hit when loading brushes with the old-style (non-json) format - a change that was made ~8 years ago.

Does this mean that mypaint still ships with old-style brushes, or that I have old-style brushes somewhere on my system that it is finding? If the latter, where, and what should I do about it? I don't actually use mypaint; I just package it in MacPorts and open it occasionally to verify that it opens.

**Comment by jplloyd** [comment]
It does not ship with old-style brushes, but still supports them (or should support them). Looks like you have one or more old-style brushes in your brush history.

Deleting/moving the files `/Users/rschmidt/.local/share/mypaint/brushes/history*` and `/Users/rschmidt/.config/mypaint/settings.json` should make this error go away on your end (still needs to be fixed, which I will try to do this evening).

You also have an old-style configuration directory on your system (changed ~7 ⅓ years ago) that you may want to remove - though it shouldn't have had any direct impact on this.

**Comment by ryandesign** [comment]
> Deleting/moving the files `/Users/rschmidt/.local/share/mypaint/brushes/history*` and `/Users/rschmidt/.config/mypaint/settings.json` should make this error go away on your end (still needs to be fixed, which I will try to do this evening).

I didn't have those paths, but I did have ~/.mypaint/brushes/history* and ~/.mypaint/settings.json, and trashing them did make the message go away; thanks.

**Comment by jplloyd** [comment]
Fixed in: 8e849a61fdf77f5958749b0831f36c4fdaf1419c

Turns out this would indeed only be triggered by ancient brush history files on Py3 builds. Nevertheless, I got to cut out some dead code, so I'm happy.
